ydb/src/client_topic/topicwriter/writer.rs

```diff
@@ -0,0 +1,405 @@
+                        return ();
+                    }
+                }
+                if writer_loop_cancellation_token.is_cancelled() {
+                    break;
+                }
+            }
+        });
+        let mut receive_messages_loop = tokio::spawn(async move {
+            let mut stream = stream; // force move inside
+            let mut reception_queue = message_loop_reception_queue; // force move inside
+
+            loop {
+                tokio::select! {
+                    _ = message_receive_loop_cancellation_token.cancelled() => { return (); }
+                    message_receive_it_res = TopicWriter::receive_messages_loop_iteration(
+                                                          stream.borrow_mut(),
+                                                          reception_queue.borrow_mut()) => {
+                        match message_receive_it_res {
+                            Ok(_) => {}
+                            Err(receive_message_iteration_error) => {
+                                message_receive_loop_cancellation_token.cancel();
+                                let mut writer_state =
+                                    writer_state_ref_message_receive_loop.lock().unwrap(); // TODO handle error
+                                *writer_state =
+                                    TopicWriterState::FinishedWithError(receive_message_iteration_error);
+                                return ();
+                            }
+                        }
+                    }
+                }
+            }
+        });
+
+        Ok(Self {
+            path: writer_options.topic_path.clone(),
+            producer_id: writer_options.producer_id.clone(),
+            partition_id: init_response.partition_id,
+            session_id: init_response.session_id,
+            last_seq_num_handled: init_response.last_seq_no,
+            write_request_messages_chunk_size: writer_options
+                .write_request_messages_chunk_size
+                .clone(),
+            write_request_send_messages_period: writer_options
+                .write_request_send_messages_period
+                .clone(),
+            auto_set_seq_no: writer_options.auto_seq_no.clone(),
+            codecs_from_server: RawSupportedCodecs::from(init_response.supported_codecs),
+            writer_message_sender: messages_sender,
+            writer_loop,
+            receive_messages_loop,
+            cancellation_token,
+            writer_state: topic_writer_state,
+            confirmation_reception_queue: Arc::new(Mutex::new(TopicWriterReceptionQueue::new())),
+            connection_manager,
+        })
+    }
+
+    async fn write_loop_iteration(
+        messages_receiver: &mut Receiver<TopicWriterMessage>,
+        task_params: &WriterPeriodicTaskParams,
+    ) -> YdbResult<()> {
+        let mut start = Instant::now();
+        let mut messages = vec![];
+        let elapsed = start.elapsed();
```

Comment:
In the line "elapsed" is fixed value, equals for time duration of init messages vector. Rest code used the duration.

It look as mistake - for check timeout code must get new elapsed on every loop.